### PR TITLE
Implemented player entity interaction

### PR DESCRIPTION
**Pull request by Spajker7**
## Introduction
This pull request implements basic changes needed for Player interaction with entities.

### Relevant issues
/

### API changes
Adds a new event: PlayerGentlyTouchEntityEvent
Adds a new method Entity::interact()

### Behavioural changes
/

## Backwards compatibility
The changes are backwards compatible.

## Follow-up
Rename the event to something sensible. 

## Tests
/

### CODE REVIEW

File: src/pocketmine/Player.php [hunk 1]
```diff
@@ -2241,7 +2242,34 @@ public function handleInventoryTransaction(InventoryTransactionPacket $packet) :
...
 				switch($type){
 					case InventoryTransactionPacket::USE_ITEM_ON_ENTITY_ACTION_INTERACT:
-						break; //TODO
+						if(!$target->isAlive()){
+							return true;
+						}
+
+						$heldItem = $this->inventory->getItemInHand();
+
+						$ev = new PlayerGentlyTouchEntityEvent($this, $target, $heldItem, $packet->trData->clickPos);
```
Comment (dktapps) [inline src/pocketmine/Player.php]: I hope you don't expect me to merge this...

File: src/pocketmine/Player.php [hunk 2]
```diff
@@ -2241,7 +2242,34 @@ public function handleInventoryTransaction(InventoryTransactionPacket $packet) :
...
+
+						if(!$this->canInteract($target, 8)){
+							$ev->setCancelled();
+						}
+
+						$this->server->getPluginManager()->callEvent($ev);
+
+						$target->interact($ev);
```
Comment (dktapps) [inline src/pocketmine/Player.php]: what exactly is the point of passing the event here? Seems like extra work for no good reason.

File: src/pocketmine/event/player/PlayerGentlyTouchEntityEvent.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		return $this->item;
+	}
+
+	/**
+	 * @return Vector3
+	 */
+	public function getClickPos(): Vector3{
+		return $this->clickPos;
+	}
+}
```
Comment (dktapps) [inline src/pocketmine/event/player/PlayerGentlyTouchEntityEvent.php]: formatting needs fixing throughout this file

File: src/pocketmine/entity/Entity.php [hunk 4]
```diff
@@ -898,6 +899,13 @@ public function attack(EntityDamageEvent $source) : void{
 		$this->setHealth($this->getHealth() - $source->getFinalDamage());
 	}
 
+	/**
+	 * @param PlayerGentlyTouchEntityEvent $source
+	 */
+	public function interact(PlayerGentlyTouchEntityEvent $source) : void{
```
Comment (dktapps) [inline src/pocketmine/entity/Entity.php]: Please let's not have another godawful mess like `Entity->attack()`. The intent of this function is not clear from its name. Is the entity interacting with some other entity, or is it the one getting interacted with?

File: src/pocketmine/Player.php [hunk 5]
```diff
@@ -2241,7 +2242,34 @@ public function handleInventoryTransaction(InventoryTransactionPacket $packet) :
...
+
+						if(!$ev->isCancelled()){
+							$oldItem = clone $heldItem;
+
+							if($heldItem->onClickEntity($this, $target, $packet->trData->clickPos)){
+								if(!$heldItem->equalsExact($oldItem)){
+									$this->inventory->setItemInHand($heldItem);
+									$this->inventory->sendHeldItem($this->hasSpawned);
```
Comment (dktapps) [inline src/pocketmine/Player.php]: I'm surprised this isn't automatic... that's perhaps a bug.

File: src/pocketmine/Player.php [hunk 6]
```diff
@@ -2241,7 +2242,34 @@ public function handleInventoryTransaction(InventoryTransactionPacket $packet) :
...
+
+						if(!$this->canInteract($target, 8)){
+							$ev->setCancelled();
+						}
+
+						$this->server->getPluginManager()->callEvent($ev);
+
+						$target->interact($ev);
+
+						if(!$ev->isCancelled()){
```
Comment (dktapps) [inline src/pocketmine/Player.php]: Seems to be an inconsistency here. The item is only receiving the click event if the event is not cancelled, but the entity will always receive it, cancelled or not.

Additionally, parameter inconsistency exists here.

**Review by dktapps — CHANGES_REQUESTED** [review]
I've given this a quick once-over, but it's not in depth.

**Comment by dktapps** [comment]
Note also that a similar change set is also present in #2124 .

File: src/pocketmine/Player.php (hunk 6 above)
Comment (Spajker7) [reply src/pocketmine/Player.php]: For Entity::interact() I tried to be consistent with other methods in Entity ( Entity::attack ). 
For Item::onClickEntity, I tried to be consistent with other methods in Item ( Item::onClickAir, Item::onActivate ).

File: src/pocketmine/event/player/PlayerGentlyTouchEntityEvent.php (hunk 3 above)
Comment (Spajker7) [reply src/pocketmine/event/player/PlayerGentlyTouchEntityEvent.php]: Whats wrong exactly ?

Comment (Muqsit) [reply src/pocketmine/event/player/PlayerGentlyTouchEntityEvent.php]: leave space after the colon

File: src/pocketmine/Player.php (hunk 2 above)
Comment (dktapps) [reply src/pocketmine/Player.php]: Entity::attack() calls the event passed to it. Here you're calling it externally and then passing it to the hook, which seems pointless.

**Comment by iiFlamiinBlaze** [comment]
#2124 ...

**Comment by Spajker7** [comment]
@iiFlamiinBlaze What's your point?

**Comment by Muqsit** [comment]
@iiFlamiinBlaze Better? I don't see how any of the PRs are better than the other (neither have you stated any ways one is better over the other), and if you think so, please recommend what needs to be changed rather than starting such silly fights.

The other PR has been inactive for more than 2 months, so why not let someone else implement it rather than waiting an eternity?

**Comment by iiFlamiinBlaze** [comment]
@Muqsit I was not starting a fight, I was just stating on the way it was implemented, that it is better. I also asked why he decided to make this PR when obviously there is one already open with the same exact feature. I get that the PR is inactive, but its still open so there must be some significance to keep it open.

**Comment by Spajker7** [comment]
Again, you haven't explained how the other PR is better? I looked at the code and it seems to be the exact thing I did, with the exception of my implementation also having a method inside the Item class.

**Pull request closed without merging**